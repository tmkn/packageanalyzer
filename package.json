{
  "name": "@tmkn/packageanalyzer",
  "version": "0.9.0-beta.3",
  "description": "A tool to analyze a Node.js project/package",
  "main": "./build/src/index.js",
  "types": "./build/src/index.d.ts",
  "files": [
    "/build/src/**/*.js",
    "/build/src/**/*.d.ts"
  ],
  "bin": {
    "pa": "./build/src/cli.js"
  },
  "scripts": {
    "build": "tsc",
    "dev": "tsc --watch",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:coverage": "jest --coverage",
    "test:clean": "node build/tests/clean.js",
    "prettier": "prettier --write \"{src,tests,utility}/**/{*.ts,*.js}\"",
    "pa": "node build/src/cli.js",
    "preview": "npm pack && tar -xvzf *.tgz && rm -rf package *.tgz",
    "lint": "eslint \"src/**\" \"tests/**\"",
    "lint:fix": "eslint \"src/**\" \"tests/**\" --fix",
    "publish": "npm publish --access public --tag beta",
    "publish:test": "npm publish --dry-run"
  },
  "author": "tmkn",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tmkn/packageanalyzer.git"
  },
  "keywords": [
    "package",
    "analyzer",
    "dependency",
    "tree"
  ],
  "bugs": {
    "url": "https://github.com/tmkn/packageanalyzer/issues"
  },
  "devDependencies": {
    "@types/express": "^4.17.2",
    "@types/jest": "^24.0.25",
    "@types/node": "^13.1.1",
    "@types/semver": "^6.0.0",
    "@typescript-eslint/eslint-plugin": "^2.13.0",
    "@typescript-eslint/parser": "^2.13.0",
    "eslint": "^6.8.0",
    "express": "^4.17.1",
    "jest": "^24.9.0",
    "prettier": "^1.19.1",
    "ts-jest": "^24.2.0",
    "typescript": "3.7.4"
  },
  "dependencies": {
    "dayjs": "^1.8.18",
    "ora": "^4.0.3",
    "semver": "^7.1.1"
  }
}
