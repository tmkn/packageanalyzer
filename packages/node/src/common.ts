import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";

import dayjs from "dayjs";

import type { DependencyTypes } from "../../shared/src/reports/Validation.js";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        // adjust when moving this function to another folder
        const file = path.join(__dirname, "..", "..", "..", "package.json");
        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}
