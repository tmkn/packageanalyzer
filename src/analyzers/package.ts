import { Writable } from "stream";
import { ExtensionData, IDataExtension, IStaticDataExtension } from "../extensions/extension";
import { INpmPackageVersion, IMalformedLicenseField } from "../npm";
import { ITreeFormatter, print } from "../tree";

type Name = string;
type Version = string;
type License = string;

export type LicenseSummary = Map<Name, Map<Version, License>>;
export type GroupedLicenseSummary = Array<{ license: string; names: string[] }>;
export type VersionSummary = Map<Name, Set<Version>>;

interface IDeprecatedInfo {
    deprecated: boolean;
    message: string;
}

interface IPackageStatistics {
    all: Package[];
    transitiveDependenciesCount: number;
    distinctByNameCount: number;
    distinctByVersionCount: number;
    path: Array<[string, string]>;
    mostReferred: [string, number];
    mostDependencies: Package;
    mostVersions: VersionSummary;
    loops: Package[];
    licenses: LicenseSummary;
    licensesByGroup: GroupedLicenseSummary;
    timeSpan: number | undefined;
    size: number | undefined;
    directDependencies: Package[];
    printDependencyTree(stdout: Writable): void;
    getExtensionData<T extends IStaticDataExtension<any, []>>(extension: T): ExtensionData<T>;
    addExtensionData(extension: IDataExtension<any>): Promise<void>;
}

export class Package implements IPackageStatistics {
    parent: Package | null = null;
    isLoop = false;
    private _extensionData: Map<Symbol, any> = new Map();
    private readonly _dependencies: Package[] = [];

    constructor(private readonly _data: Readonly<INpmPackageVersion>) {}

    get name(): string {
        return this._data.name;
    }

    get version(): string {
        return this._data.version;
    }

    get fullName(): string {
        return `${this.name}@${this.version}`;
    }

    get timeSpan(): number | undefined {
        throw new Error("Not Implemented");
    }

    get size(): number | undefined {
        throw new Error("Not Implemented");
    }

    get license(): string {
        try {
            const { license } = this._data;

            //check if license field is set
            if (typeof license !== "undefined") {
                if (typeof license === "string") {
                    return license;
                } else if (this._isLicenseObject(license)) {
                    return license.type;
                } else {
                    return JSON.stringify(license);
                }
                //fallback to licenses field
            } else if (typeof this._data.licenses !== "undefined")
                return this._data.licenses.map(l => l.type).join(",");
            //weird format | not set -> fail
            else {
                throw new Error(`Unable to parse license`);
            }
        } catch {
            return `PARSE ERROR: ${this.fullName}`;
        }
    }

    //even though license should be string some packages contain json objects...
    private _isLicenseObject(data: unknown): data is IMalformedLicenseField {
        if (typeof data === "object" && data !== null) {
            return "type" in data && "url" in data;
        }

        return false;
    }

    get directDependencyCount(): number {
        return this._dependencies.length;
    }

    get deprecatedInfo(): IDeprecatedInfo {
        const { deprecated } = this._data;

        if (deprecated) {
            return {
                deprecated: true,
                message: deprecated
            };
        }

        return {
            deprecated: false,
            message: ``
        };
    }

    getData<T extends keyof INpmPackageVersion = keyof INpmPackageVersion>(
        key: T
    ): INpmPackageVersion[T] {
        return this._data[key];
    }

    addDependency(dependency: Package): void {
        dependency.parent = this;

        this._dependencies.push(dependency);
    }

    visit(
        callback: (dependency: Package) => void,
        includeSelf = false,
        start: Package = this
    ): void {
        if (includeSelf) callback(this);

        for (const child of start._dependencies) {
            callback(child);
            this.visit(callback, false, child);
        }
    }

    getPackagesBy(filter: (pkg: Package) => boolean): Package[] {
        const matches: Package[] = [];

        this.visit(
            d => {
                if (filter(d)) matches.push(d);
            },
            true,
            this
        );

        return matches;
    }

    getPackagesByName(name: string, version?: string): Package[] {
        const matches: Package[] = [];

        this.visit(d => {
            if (typeof version === "undefined") {
                if (d.name === name) matches.push(d);
            } else {
                if (d.name === name && d.version === version) matches.push(d);
            }
        }, true);

        return matches;
    }

    getPackageByName(name: string, version?: string): Package | null {
        const matches: Package[] = this.getPackagesByName(name, version);

        if (matches.length > 0) return matches[0];

        return null;
    }

    get transitiveDependenciesCount(): number {
        let count = 0;

        this.visit(() => count++);

        return count;
    }

    get licenses(): LicenseSummary {
        const licenseMap: LicenseSummary = new Map();

        this.visit(d => {
            const packageKey = licenseMap.get(d.name);

            if (!packageKey) {
                licenseMap.set(d.name, new Map([[d.version, d.license]]));
            } else {
                packageKey.set(d.version, d.license);
            }
        }, true);

        return licenseMap;
    }

    get licensesByGroup(): GroupedLicenseSummary {
        const licenses = this.licenses;
        const sorted: Map<string, Set<string>> = new Map();
        const grouped: GroupedLicenseSummary = [];

        for (const [name, versions] of licenses) {
            for (const license of versions.values()) {
                const entry = sorted.get(license);

                if (entry) {
                    entry.add(name);
                } else {
                    sorted.set(license, new Set([name]));
                }
            }
        }

        for (const [license, names] of sorted) {
            grouped.push({
                license: license,
                names: [...new Set([...names.values()].sort())]
            });
        }

        return grouped.sort((a, b) => b.names.length - a.names.length);
    }

    get sorted(): Map<string, Map<string, Package>> {
        const sorted: Map<string, Map<string, Package>> = new Map();

        this.visit(d => {
            const packageName = sorted.get(d.name);

            if (packageName) {
                packageName.set(d.version, d);
            } else {
                sorted.set(d.name, new Map([[d.version, d]]));
            }
        }, true);

        return sorted;
    }

    //todo possible multiple matches
    get mostDependencies(): Package {
        let most: Package = this;

        this.visit(d => {
            if (most._dependencies.length < d._dependencies.length) {
                most = d;
            }
        });

        return most;
    }

    //todo possible multiple matches
    get mostReferred(): [string, number] {
        const mostReferred: Map<string, number> = new Map();
        let max = 0;
        let name = "";

        this.visit(d => {
            const entry = mostReferred.get(d.name);

            if (entry) {
                mostReferred.set(d.name, entry + 1);
            } else {
                mostReferred.set(d.name, 1);
            }
        }, true);

        for (const [pkgName, count] of mostReferred) {
            if (count > max) {
                max = count;
                name = pkgName;
            }
        }

        return [name, max];
    }

    get mostVersions(): VersionSummary {
        let max = 0;
        let map: VersionSummary = new Map();

        for (const [name, versions] of this.sorted) {
            if (versions.size > max) {
                max = versions.size;
                map = new Map([[name, new Set(versions.keys())]]);
            } else if (max === versions.size) {
                map.set(name, new Set(versions.keys()));
            }
        }

        return map;
    }

    get cost(): number {
        let cost = 0;

        this.visit(d => {
            const dist = d.getData("dist");

            if (dist?.unpackedSize) {
                cost += dist.unpackedSize;
            }
        }, true);

        return cost;
    }

    get path(): Array<[string, string]> {
        const path: Array<[string, string]> = [];
        let current: Package | null = this;

        while (current.parent !== null) {
            path.push([current.name, current.version]);

            current = current.parent;
        }

        path.push([current.name, current.version]);

        return path.reverse();
    }

    get pathString(): string {
        const levels: string[] = [];

        for (const [name, version] of this.path) {
            levels.push(`${name}@${version}`);
        }

        return levels.join(" â†’ ");
    }

    get all(): Package[] {
        const all: Package[] = [];

        this.visit(d => all.push(d), true);

        return all;
    }

    get loops(): Package[] {
        const loops: Package[] = [];

        this.visit(d => {
            if (d.isLoop) loops.push(d);
        }, true);

        return loops;
    }

    //returns the loop path e.g. c->d->c instead of the whole path a->b->c->d->c
    get loopPathString(): string {
        const split = this.pathString.indexOf(this.fullName);

        return this.pathString.slice(split);
    }

    get loopPathMap(): ReadonlyMap<string, Set<string>> {
        const map: Map<string, Set<string>> = new Map();
        const loops = this.loops;

        for (const loop of loops) {
            const entry = map.get(loop.name);

            if (entry) {
                entry.add(loop.loopPathString);
            } else {
                map.set(loop.name, new Set([loop.loopPathString]));
            }
        }

        return map;
    }

    get distinctLoopCount(): number {
        return [...this.loopPathMap].reduce((i, [, loops]) => i + loops.size, 0);
    }

    get distinctByNameCount(): number {
        const packageNames: Set<string> = new Set();

        this.visit(d => packageNames.add(d.name));

        return packageNames.size;
    }

    get distinctByVersionCount(): number {
        const packageNames: Set<string> = new Set();

        this.visit(d => packageNames.add(d.fullName));

        return packageNames.size;
    }

    get directDependencies(): Package[] {
        const depth = this.path.length;

        return this.getPackagesBy(pkg => pkg.path.length === depth + 1);
    }

    printDependencyTree(stdout: Writable): void {
        const converter: ITreeFormatter<Package> = {
            getLabel: data => `${data.fullName} (${data.transitiveDependenciesCount} dependencies)`,
            getChildren: data => data.directDependencies
        };

        print<Package>(this, converter, stdout);
    }

    getExtensionData<T extends IStaticDataExtension<any, any[]>>(extension: T): ExtensionData<T> {
        const data = this._extensionData.get(extension.key);

        if (typeof data === "undefined") {
            throw new Error(`No extension data found for ${extension.toString()}`);
        }

        return data;
    }

    async addExtensionData(extension: IDataExtension<any>): Promise<void> {
        const data = await extension.apply(this);

        this._extensionData.set(extension.key, data);
    }
}
