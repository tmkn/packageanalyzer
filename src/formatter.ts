import { Writable } from "stream";

type LineKeyValue = [string, string];

export interface IFormatter {
    writeLine: (line: string) => void;
    writeGroup: (lines: Array<LineKeyValue | string>) => void;
}

export class Formatter implements IFormatter {
    constructor(private _writer: Writable) {}

    writeLine(line: string): void {
        this._writer.write(`${line}\n`);
    }

    writeGroup(lines: Array<LineKeyValue | string>): void {
        const padding: number =
            lines.reduce((prev, current) => {
                if (Array.isArray(current)) {
                    const [key] = current;

                    if (key.length > prev) return key.length;
                }

                return prev;
            }, 0) + 2;

        for (const line of lines) {
            if (typeof line === "string") {
                this.writeLine(line);
            } else {
                const [key, value] = line;

                this.writeLine(`${`${key}:`.padEnd(padding)}${value}`);
            }
        }
    }
}
