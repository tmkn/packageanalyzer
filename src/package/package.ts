import { get } from "lodash";
import { DecoratorData, DecoratorKey, IDecorator } from "../extensions/decorators/Decorator";

import { IPackageJson } from "../npm";

interface IDeprecatedInfo {
    deprecated: boolean;
    message: string;
}

interface IPackage<T> {
    parent: T | null;
    isLoop: boolean;
    name: string;
    version: string;
    fullName: string;
    directDependencies: T[];
    deprecatedInfo: IDeprecatedInfo;

    addDependency: (dependency: T) => void;

    visit: (callback: (dependency: T) => void, includeSelf: boolean) => void;
    getPackagesBy: (filter: (pkg: T) => boolean) => T[];
    getPackagesByName: (name: string, version?: string) => T[];
    getPackageByName: (name: string, version?: string) => T | null;

    getData(key: string): unknown;

    getDecoratorData<D extends IDecorator<any, unknown>>(key: DecoratorKey<D>): DecoratorData<D>;
    setDecoratorData<D extends IDecorator<any, unknown>>(
        key: DecoratorKey<D>,
        data: DecoratorData<D>
    ): void;
}

export class Package implements IPackage<Package> {
    parent: Package | null = null;
    isLoop = false;

    private _decoratorData: Map<IDecorator<any, any>, any> = new Map();
    private readonly _dependencies: Package[] = [];

    constructor(private readonly _data: Readonly<IPackageJson>) {}

    get name(): string {
        return this._data.name;
    }

    get version(): string {
        return this._data.version;
    }

    get fullName(): string {
        return `${this.name}@${this.version}`;
    }

    get directDependencies(): Package[] {
        return this._dependencies;
    }

    get deprecatedInfo(): IDeprecatedInfo {
        const deprecated = this.getData("deprecated");

        if (typeof deprecated === "string") {
            return {
                deprecated: true,
                message: deprecated
            };
        }

        return {
            deprecated: false,
            message: ``
        };
    }

    addDependency(dependency: Package): void {
        dependency.parent = this;

        this._dependencies.push(dependency);
    }

    visit(callback: (dependency: Package) => void, includeSelf = false): void {
        if (includeSelf) callback(this);

        for (const child of this._dependencies) {
            callback(child);
            child.visit(callback, false);
        }
    }

    getPackagesBy(filter: (pkg: Package) => boolean): Package[] {
        const matches: Package[] = [];

        this.visit(d => {
            if (filter(d)) matches.push(d);
        }, true);

        return matches;
    }

    getPackagesByName(name: string, version?: string): Package[] {
        const matches: Package[] = [];

        this.visit(d => {
            if (typeof version === "undefined") {
                if (d.name === name) matches.push(d);
            } else {
                if (d.name === name && d.version === version) matches.push(d);
            }
        }, true);

        return matches;
    }

    getPackageByName(name: string, version?: string): Package | null {
        const matches: Package[] = this.getPackagesByName(name, version);

        return matches[0] ?? null;
    }

    getData(key: string): unknown {
        return get(this._data, key);
    }

    getDecoratorData<E extends IDecorator<any, unknown>>(key: DecoratorKey<E>): DecoratorData<E> {
        const data = this._decoratorData.get(key);

        if (typeof data === "undefined") {
            throw new Error(`No extension data found for ${key.toString()}`);
        }

        return data;
    }

    setDecoratorData<E extends IDecorator<any, unknown>>(
        key: DecoratorKey<E>,
        data: DecoratorData<E>
    ): void {
        this._decoratorData.set(key, data);
    }
}
