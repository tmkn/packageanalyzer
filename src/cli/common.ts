import * as path from "path";
import * as fs from "fs";

import * as dayjs from "dayjs";
import * as chalk from "chalk";

import { PackageAnalytics, VersionSummary, GroupedLicenseSummary } from "../analyzers/package";
import { DependencyTypes } from "../visitors/visitor";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const file = path.join(__dirname, "./../../package.json");

        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch (e) {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

export function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(chalk.bold(`Statistics for ${pa.fullName}\n`));

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa, padding, paddingLeft);
    printLicenseInfo(pa.licensesByGroup, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest package:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
        console.log(`${`Newest package path:`.padEnd(padding)}${newest.pathString}`);
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest package:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
        console.log(`${`Oldest package path:`.padEnd(padding)}${oldest.pathString}`);
    }
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(pa: PackageAnalytics, padding: number, paddingLeft: number): void {
    const { loops, loopPathMap, distinctLoopCount } = pa;

    console.log(`${`Loops:`.padEnd(padding)}${loops.length} (${distinctLoopCount} distinct)`);

    if (distinctLoopCount > 0) {
        const [first] = loops.map(l => l.pathString).sort();

        console.log(`    affected Packages: [${[...loopPathMap.keys()].join(", ")}]`);
        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(groupedLicenses: GroupedLicenseSummary, paddingLeft: number): void {
    console.log(`Licenses:`);
    for (const { license, names } of groupedLicenses) {
        const threshold = 4;
        const samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${
                    names.length - threshold
                } more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }
}
