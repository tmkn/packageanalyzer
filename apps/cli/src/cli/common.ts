import { Command } from "clipanion";

import { AbstractReport } from "../../../../packages/shared/src/reports/Report.js";
import { ReportService } from "../../../../packages/shared/src/reports/ReportService.js";
import { Formatter, type IFormatter } from "../../../../packages/shared/src/utils/formatter.js";
import { NodeHost } from "../../../../packages/node/src/host/NodeHost.js";

export abstract class CliCommand<
    T extends AbstractReport<any> | AbstractReport<any>[]
> extends Command {
    abstract getReports(): T | Promise<T>;

    exitCode = 0;

    beforeProcess: ((report: T) => void) | undefined = undefined;

    async execute(): Promise<number | void> {
        const host = new NodeHost(this.context.stdout, this.context.stderr);

        try {
            const reports = await this.getReports();
            const reportService = new ReportService(
                {
                    reports: Array.isArray(reports) ? reports : [reports]
                },
                host
            );

            this.beforeProcess?.(reports);
            this.exitCode = (await reportService.process()) ?? 0;
        } catch (e: unknown) {
            const stderrFormatter: IFormatter = new Formatter(host.getStderrWriter());

            if (e) stderrFormatter.writeLine(e?.toString());
            this.exitCode = 1;
        } finally {
            await Promise.all([host.getStdoutWriter().flush(), host.getStderrWriter().flush()]);
        }

        return this.exitCode;
    }
}
