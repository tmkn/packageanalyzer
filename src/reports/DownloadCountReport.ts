import * as t from "io-ts";

import { INpmDownloadStatistic } from "../npm";
import { Package } from "../package/package";
import { IFormatter } from "../utils/formatter";
import { downloadJson, Url, urlType } from "../utils/requests";
import { getPackageVersionfromString, PackageVersion } from "../visitors/visitor";
import { AbstractReport, IReportContext } from "./Report";
import { BasePackageParameter } from "./Validation";

const OptionalParams = t.partial({
    url: urlType
});

const DownloadParams = t.intersection([BasePackageParameter, OptionalParams]);

export type IDownloadParams = t.TypeOf<typeof DownloadParams>;

export class DownloadReport extends AbstractReport<IDownloadParams> {
    name = `Download Report`;
    pkg: PackageVersion;

    constructor(params: IDownloadParams) {
        super(params);

        this.depth = 0;

        this.pkg = getPackageVersionfromString(params.package);
    }

    async report({ stdoutFormatter }: IReportContext, pkg: Package): Promise<void> {
        await cliDownloads(pkg.name, this.params.url ?? null, stdoutFormatter);
    }

    override validate(): t.Type<IDownloadParams> {
        return DownloadParams;
    }
}

async function cliDownloads(pkg: string, url: Url | null, formatter: IFormatter): Promise<void> {
    try {
        const downloads =
            url !== null ? await getDownloadsLastWeek(pkg, url) : await getDownloadsLastWeek(pkg);

        formatter.writeLine(`${pkg}: ${downloads.downloads} Downloads`);
    } catch (e) {
        console.log(e);
        formatter.writeLine(`Couldn't get downloads for ${pkg}`);
    }
}

export async function getDownloadsLastWeek(
    name: string,
    url: Url = `https://api.npmjs.org/downloads/point/last-week/`
): Promise<INpmDownloadStatistic> {
    const json = await downloadJson<INpmDownloadStatistic>(`${url}${encodeURIComponent(name)}`);

    if (json !== null) return json;

    throw new Error(`Couldn't get download numbers for ${name}`);
}
