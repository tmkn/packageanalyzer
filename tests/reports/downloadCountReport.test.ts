import { Package } from "../../src/package/package";
import { DownloadReport } from "../../src/reports/DownloadCountReport";
import { Formatter } from "../../src/utils/formatter";
import { TestWritable } from "./../common";
import { createMockDownloadServer, IMockServer } from "./../server";

describe(`DownloadCountReport Tests`, () => {
    let server: IMockServer;

    beforeAll(async () => {
        server = await createMockDownloadServer();
    });

    afterAll(() => server.close());

    test(`works`, async () => {
        const downloadReport = new DownloadReport({
            package: `_downloads`,
            url: `http://localhost:${server.port}/`
        });

        //@ts-expect-error
        const fakePgk: Package = {
            name: `_downloads`
        };
        const stdout = new TestWritable();
        const stdoutFormatter = new Formatter(stdout);
        const stderr = new TestWritable();
        const stderrFormatter = new Formatter(stderr);

        await downloadReport.report(fakePgk, { stdoutFormatter, stderrFormatter });

        expect(stdout.lines).toMatchSnapshot(`stdout`);
        expect(stderr.lines).toMatchSnapshot(`stderr`);
    });
});
