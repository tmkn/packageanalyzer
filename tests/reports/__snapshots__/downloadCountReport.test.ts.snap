// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadCountReport Tests works: stderr 1`] = `Array []`;

exports[`DownloadCountReport Tests works: stdout 1`] = `
Array [
  "_downloads: 8609192 Downloads",
  "",
]
`;
